docs/interop-ksp.md

```diff
@@ -0,0 +1,126 @@
+[Kotlin Symbol Processing](https://github.com/google/ksp) (KSP) types to KotlinPoet types and
+writing to KSP `CodeGenerator`.
+
+Note that this API is currently in preview and subject to API changes. Usage of them requires opting
+in to the `@KotlinPoetKspPreview` annotation.
+
+### Examples
+
+Examples are based on reading the following property as a `KSProperty`
+
+```kotlin
+class Taco {
+  internal inline val seasoning: String get() = "spicy"
+}
+```
+
+**Convert a `KSType` to a `TypeName`**
+
+```kotlin
+// returns a `ClassName` of value `kotlin.String`
+seasoningKsProperty.type.toTypeName()
+```
+
+**Convert a `Modifier` to a `KModifier`**
+
+```kotlin
+// returns `[KModifier.INLINE]`
+seasoningKsProperty.modifiers.mapNotNull { it.toKModifier() }
+```
+
+**Convert a `Visibility` to a `KModifier`**
+
+```kotlin
+// returns `KModifier.INTERNAL`
+seasoningKsProperty.getVisibility().toKModifier()
+```
+
+**Write to `CodeGenerator`**
+
+To write a `FileSpec` to a KSP `CodeGenerator`, simply call the `FileSpec.writeTo(CodeGenerator, ...)`
+extension function.
+
+```kotlin
+fileSpec.writeTo(codeGenerator)
+```
+
+### Type Parameters
+
+Type parameters can be declared on classes, functions, and typealiases. These parameters are then
+available to all of its enclosed elements. In order for these elements to resolve these in KSP, you
+must be able to reference these type parameters by their _index_.
+
+In `kotlinpoet-ksp` this is orchestrated by the `TypeParameterResolver` API, which can be passed
+into most `toTypeName()` (or similar) functions to give them access to enclosing type parameters
+that they may reference.
+
+The canonical way to create an instance of this is to call `toTypeParameterResolver()` on a
+`List<toTypeParameterResolver>`.
+
+Consider the following class and function
+
+```kotlin
+abstract class Taco<T> {
+  abstract val seasoning: T
+}
+```
+
+To properly resolve the type of `seasoning`, we need to pass the class `TypeParameterResolver` to
+`toTypeName()` so that it can properly resolve it.
+
+```kotlin
+val classTypeParams = ksClassDeclaration.typeParameters.toTypeParameterResolver()
+// returns `T`
+val seasoningType = seasoningKsProperty.type.toTypeName(classTypeParams)
+```
+
+`TypeParameterResolver` is also composable to allow for multi-level nesting. `toTypeParameterResolver()`
+has an optional `parent` parameter to provide a parent instance.
+
+Consider our previous example again, but this time with a function that defines its own type parameters.
+
+```kotlin
+class Taco<T> {
+  fun <E> getShellOfType(param1: E, param2: T) {
+
+  }
+}
+```
+
+To resolve its parameters, we need to create a `TypeParameterResolver` from the function's
+`typeParameters` and _compose_ it with the enclosing class's type parameters as a `parent`.
+
+```kotlin
+val classTypeParams = ksClassDeclaration.typeParameters.toTypeParameterResolver()
+val functionTypeParams = ksFunction.typeParameters.toTypeParameterResolver(parent = classTypeParams)
```

Comment:
I like this API!

Reply:
yeah it's worked well. We use basically the same thing in metadata, so I wonder if it would be worth trying to share somewhere.